pkg/lib.go

```diff
@@ -378,21 +367,95 @@ func GetInventoryReports(cfg *config.Application) (AccountRoutedReports, error)
-				if err != nil {
-					return AccountRoutedReports{}, err
+			accountReport, err := GetInventoryReportForNamespaces(cfg, namespaces)
+			if err != nil {
+				return BatchedReports{}, err
+			}
+			reports[account] = accountReport
+		}
+	}
+
+	return getBatchedInventoryReports(reports, cfg.InventoryReportLimits.Namespaces), nil
+}
+
+//nolint:gocognit,funlen
+func getBatchedInventoryReports(reports AccountRoutedReports, batchSize int) BatchedReports {
+	batchedReports := BatchedReports{}
+	if batchSize <= 0 {
+		for account, report := range reports {
+			batchedReports[account] = append(batchedReports[account], report)
+		}
+		return batchedReports
+	}
+
+	log.Infof("Batching namespaces into groups of %d", batchSize)
+	for account, accountReport := range reports {
+		if len(accountReport.Namespaces) <= batchSize {
+			batchedReports[account] = append(batchedReports[account], accountReport)
+			continue
+		}
+		namespaceBatches := make([][]inventory.Namespace, 0)
+		// Construct batches of namespaces
+		for i := 0; i < len(accountReport.Namespaces); i += batchSize {
+			end := i + batchSize
+			if end > len(accountReport.Namespaces) {
+				end = len(accountReport.Namespaces)
+			}
+			namespaceBatches = append(namespaceBatches, accountReport.Namespaces[i:end])
+		}
+
+		nodeMap := make(map[string]inventory.Node)
+		for _, node := range accountReport.Nodes {
+			nodeMap[node.UID] = node
+		}
+		podMap := make(map[string]inventory.Pod)
+		for _, pod := range accountReport.Pods {
+			podMap[pod.UID] = pod
+		}
+		containersByPod := make(map[string][]inventory.Container)
+		for _, container := range accountReport.Containers {
+			containersByPod[container.PodUID] = append(containersByPod[container.PodUID], container)
+		}
+		podsByNamespace := make(map[string][]inventory.Pod)
+		for _, pod := range accountReport.Pods {
+			podsByNamespace[pod.NamespaceUID] = append(podsByNamespace[pod.NamespaceUID], pod)
+		}
+		containersByNamespace := make(map[string][]inventory.Container)
+		for _, container := range accountReport.Containers {
+			namespaceUID := podMap[container.PodUID].NamespaceUID
+			containersByNamespace[namespaceUID] = append(containersByNamespace[namespaceUID], container)
+		}
+
+		// Construct reports for each batch
+		for _, batch := range namespaceBatches {
+			batchedPodsSlice := []inventory.Pod{}
+			batchedContainersSlice := []inventory.Container{}
+			batchedNodesMap := map[string]inventory.Node{}
+			for _, ns := range batch {
+				batchedPodsSlice = append(batchedPodsSlice, podsByNamespace[ns.UID]...)
+				batchedContainersSlice = append(batchedContainersSlice, containersByNamespace[ns.UID]...)
+				for _, pod := range batchedPodsSlice {
+					batchedNodesMap[pod.NodeUID] = nodeMap[pod.NodeUID]
 				}
-				reports[account] = append(reports[account], accountReport)
 			}
+			batchedNodesSlice := []inventory.Node{}
+			for _, node := range batchedNodesMap {
+				batchedNodesSlice = append(batchedNodesSlice, node)
+			}
```

Comment:
Is this just converting the map to an array? There isn't a cleaner way to do this?

Reply:
not really, there is a Values function I could call but it does this under the hood. https://cs.opensource.google/go/x/exp/+/fc45aab8:maps/maps.go;l=20

Reply:
That might be a bit cleaner but no perf gain so this is fine.